ethereum/statetransition/src/main/java/tech/pegasys/artemis/statetransition/attestation/MatchingDataAttestationGroup.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.artemis.statetransition.attestation;
+
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.NavigableMap;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+import tech.pegasys.artemis.datastructures.operations.Attestation;
+import tech.pegasys.artemis.datastructures.operations.AttestationData;
+
+class MatchingDataAttestationGroup implements Iterable<Attestation> {
```

Comment:
It might be just me but I think this class deserves an explanation through comments. Although written so well, its purpose is pretty complex so an explanation in comments would be really nice. Something would be like:
```suggestion
class MatchingDataAttestationGroup implements Iterable<Attestation> {
/*   This object keeps attestations for a single attestation data sorted in descending order by the number of set aggregation bits. Its iterator goes through these attestations with the given order and aggregates them. Aggregating by starting with attestations with the highest number of bits potentially allows us to create aggregates with highest number of bits set. */
```

Reply:
Good idea.  I've added a bunch of comments for the classes to help explain what's going on.  Generally tried to keep them at an architectural level rather than being too specific about the details though.  I'm sure the words could be better but they'll have to do...